### PR TITLE
feat: add Tracing inspector

### LINKED ISSUE #1606: Add tracing revm tracer impl
### Describe the feature

start by porting foundry tracer

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mattsse**
Closes #1606 

~~WIP~~

Opening this up for review/merge, since already very large.

This includes a new `Inspector` for recording call traces, based on anvil's inspector.

Eventually, this could perhaps be unified with anvil, but it would be that anvil uses reth's inspector because it includes some additional features.

This mainly includes a ported Trace Inspector from anvil.
some parts like, finalizing traces etc are missing and will be addressed when the RPC functions are implemented.

### CODE REVIEW

File: crates/revm/revm-inspectors/src/tracing/types.rs [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+}
+
+/// Ethereum log.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(crate) struct RawLog {
+    /// Indexed event params are represented as log topics.
+    pub(crate) topics: Vec<H256>,
+    /// Others are just plain data.
+    pub(crate) data: Bytes,
+}
```
Comment (gakonst) [inline crates/revm/revm-inspectors/src/tracing/types.rs:224]: nit - this seems like a dup type that we should be able to reuse from elsewhere

File: crates/revm/revm-inspectors/src/stack/maybe_owned.rs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    cell::{Ref, RefCell},
+    rc::Rc,
+};
+
+/// An [Inspector] that is either owned by an individual [Inspector] or is shared as part of a
+/// series of inspectors in a [InspectorStack](crate::stack::InspectorStack).
+///
+/// Caution: if the [Inspector] is _stacked_ then it _must_ be called first.
+#[derive(Debug)]
+pub enum MaybeOwnedInspector<INSP> {
```
Comment (gakonst) [inline crates/revm/revm-inspectors/src/stack/maybe_owned.rs:16]: Where do you envision this kind of inspector to be used?

File: crates/revm/revm-inspectors/src/tracing/mod.rs [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+        data: &mut EVMData<'_, DB>,
+        is_static: bool,
+    ) -> InstructionResult {
+        if self.record_steps {
+            self.gas_inspector.step(interp, data, is_static);
+            self.start_step(interp, data);
+        }
+
+        InstructionResult::Continue
+    }
```
Comment (gakonst) [inline crates/revm/revm-inspectors/src/tracing/mod.rs:258]: Hooking on step is really expensive I think even for just having an `if` branch when running a large number of transactions when `record_steps = false`..we should explore converting this into a const that we don't pay for at runtime eventually, similar to how Revm has a "const INSPECT: bool" for the entire thing

**Review by gakonst — APPROVED** [review]
good start - I'm glad we did this first in Anvil, most of the code is the same. I would keep in mind that we may want to have iterator interfaces where we can instead of vectors where possible as that may eventually unlock "streaming" traces, heavily reducing latency & peak mem usage

maybe that would work with an inspector that owns a channel `tx` and `tx.send`s to the `rx`, and the `rx` sits in the TraceApi and exposes a `impl Iterator` or `impl Stream` for Parity/Geth Traces

File: crates/revm/revm-inspectors/src/stack/maybe_owned.rs (hunk 2 above)
Comment (mattsse) [reply crates/revm/revm-inspectors/src/stack/maybe_owned.rs:16]: either only used for one inspector like tracing or as part of stack

**Comment by mattsse** [comment]
streaming items over the wire for the supported subscription protocols is only really doable for ipc, which writes directly to the socket, whereas websocket frame requires the payload length before the payload is being sent

**Comment by gakonst** [comment]
> whereas websocket frame requires the payload length before the payload is being sent

ah yikes. maybe grpc eventually

**Pull request merged**